HW1/P4/P4_bfs.ipynb

```diff
@@ -0,0 +1,74 @@
+{
+ "cells": [
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "collapsed": true
+   },
+   "outputs": [],
+   "source": [
+    "def bfs(sc, adj_matrix_rdd, char, diameter=float(\"inf\")):\n",
+    "    '''\n",
+    "    Takes SparkContext (sc), adjacency matrix rdd (adj_matrix_rdd), character name (char), and optional diameter\n",
+    "    of graph as parameter and performs BFS.\n",
+    "    Note: adj_matrix_rdd contains \n",
+    "        key = character name\n",
+    "        value = list of names of adjacent characters\n",
+    "    Note: adj_matrix_rdd should be hash partitioned into 100 parts\n",
+    "    '''\n",
+    "    def group_neighbor_dist(neighbors,dist):\n",
+    "        return [(n,dist) for n in neighbors]\n",
+    "    dist = 0\n",
+    "    \n",
+    "    num_iter_accum = sc.accumulator(0)\n",
+    "    dists_rdd = sc.parallelize([(char,dist)]).partitionBy(100) #shortest paths to nodes\n",
+    "    neighbors_rdd = dists_rdd\n",
+    "    #We assume graph diameter is <=10\n",
+    "    num_new_neighbors_accum = sc.accumulator(1)\n",
+    "    while (num_iter_accum.value < diameter) and (num_new_neighbors_accum.value != 0):\n",
```

Comment:
Well done, once again very clean easy to follow code. I walked through your BFS and had no problem understanding what each step does.